### PR TITLE
[Bugfix][310P]: fix mmencoder&swiglu cannot use via padding

### PR DESCRIPTION
**Pull request by Tflowers-0129**
### How was this patch tested?

- vLLM version: v0.15.0
- vLLM main: https://github.com/vllm-project/vllm/commit/9562912cead1f11e8540fb91306c5cbda66f0007

### CODE REVIEW

File: tests/ut/_310p/ops/test_linear_310p.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import math
```
Comment (pu-zhe) [inline tests/ut/_310p/ops/test_linear_310p.py:1]: 记得添加License信息

File: vllm_ascend/_310p/ops/activation.py [hunk 2]
```diff
@@ -16,13 +16,20 @@
...
 
 
 class AscendSiluAndMul310(AscendSiluAndMul):
     def forward(self, x: torch.Tensor) -> torch.Tensor:
-        h = x.shape[-1] // 2
-        out = (F.silu(x[..., :h].to(torch.float32)) * x[..., h:].to(torch.float32)).to(torch.float16)
+
+        orig_shape = x.shape
+        if x.dim() > 2:
+            x = x.contiguous().view(-1, orig_shape[-1])
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/activation.py:29]: 300I DUO存在这种情况？A2代码中无此逻辑，可稳定运行。

File: vllm_ascend/_310p/ops/mm_encoder_attention.py [hunk 3]
```diff
@@ -34,28 +35,50 @@ def forward_oot(
...
         bsz, q_len = query.size()[:2]
         kv_len = key.size(1)
-        query = query.view(bsz * q_len, self.num_heads, self.head_size)
-        key = key.view(bsz * kv_len, self.num_kv_heads, self.head_size)
-        value = value.view(bsz * kv_len, self.num_kv_heads, self.head_size)
+        head_size_real = int(query.shape[-1])
+
+        query = query.view(bsz * q_len, self.num_heads, head_size_real)
+        key = key.view(bsz * kv_len, self.num_kv_heads, head_size_real)
+        value = value.view(bsz * kv_len, self.num_kv_heads, head_size_real)
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/mm_encoder_attention.py:47]: head_size_real = query.shape[-1] = self.head_size。
此处为冗余更改

File: vllm_ascend/_310p/ops/mm_encoder_attention.py [hunk 4]
```diff
@@ -34,28 +35,50 @@ def forward_oot(
...
             seq_len = torch.diff(cu_seqlens.to("cpu", dtype=torch.int32))
 
+        scale = getattr(self, "scale", None)
+        if scale is None:
+            head_size_orig = getattr(self, "head_size_orig", None)
+            if head_size_orig is not None:
+                scale = float(head_size_orig) ** -0.5
+            else:
+                scale = float(head_size_real) ** -0.5
+
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/mm_encoder_attention.py:68]: 此处为冗余代码，不存在self.head_size_orig，self.scale等类属性。使用scale=self.head_size**-0.5，可覆盖pad和非pad两种场景。

File: vllm_ascend/_310p/ops/linear.py [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+        quant_config: QuantizationConfig | None = None,
+        prefix: str = "",
+        *,
+        return_bias: bool = True,
+        disable_tp: bool = False,
+    ):
+        nn.Module.__init__(self)
+
+        self.input_size = int(input_size)
+        self.output_size = int(output_size)
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/linear.py:211]: 冗余的强制类型转换

File: vllm_ascend/_310p/ops/linear.py [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+        self.tp_rank = 0 if disable_tp else get_tensor_model_parallel_rank()
+        self.tp_size = 1 if disable_tp else get_tensor_model_parallel_world_size()
+
+        self.input_size_per_partition = int(input_size)
+        self.output_size_per_partition = divide(int(output_size), int(self.tp_size))
+
+        # merged-column uses self.output_sizes; normal column => one part
+        self.output_partition_sizes = [int(self.output_size_per_partition)]
+        if hasattr(self, "output_sizes"):
+            self.output_partition_sizes = [divide(int(s), int(self.tp_size)) for s in self.output_sizes]
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/linear.py:250]: 冗余的强制类型转换。
参考vllm原仓：
```
self.input_size_per_partition = input_size
        self.output_size_per_partition = divide(output_size, self.tp_size)
        self.output_partition_sizes = [self.output_size_per_partition]
        # If QKV or MergedColumn, use output size of each partition.
        if hasattr(self, "output_sizes"):
            self.output_partition_sizes = [
                divide(output_size, self.tp_size) for output_size in self.output_sizes
            ]
```

File: vllm_ascend/_310p/ops/linear.py [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+        self.return_bias = return_bias
+        self.disable_tp = disable_tp
+
+        if quant_config is None:
+            self.quant_method: QuantizeMethodBase | None = AscendUnquantizedLinearMethod310()
+        else:
+            self.quant_method = quant_config.get_quant_method(self, prefix=prefix)
+
+
+class AscendColumnParallelLinear310(ColumnParallelLinear):
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/linear.py:225]: ColumnParallelLinear类会额外影响到qwen3-next的卷积层，glm4.1v的patch merger proj层，whisper的q_proj层等

File: vllm_ascend/_310p/ops/linear.py [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+            out = out[..., :out_real].contiguous()
+            if out_bias is not None and out_bias.numel() != out_real:
+                out_bias = out_bias[:out_real].contiguous()
+
+        if not self.return_bias:
+            return out
+        return out, out_bias
+
+
+class AscendRowParallelLinear310(RowParallelLinear):
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/linear.py:334]: RowParallelLinear类会额外影响到dense层，每个模型的down层命名不一致，有down_proj, fc2，dense_4h_to_h等，行为不可控

File: vllm_ascend/_310p/ops/linear.py [hunk 9]
```diff
@@ -0,0 +1,571 @@
...
+            from vllm.distributed import tensor_model_parallel_all_reduce
+
+            out = tensor_model_parallel_all_reduce(out)
+
+        if not self.return_bias:
+            return out
+        return out, (self.bias if self.skip_bias_add else None)
+
+
+class AscendMergedColumnParallelLinear310(MergedColumnParallelLinear):
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/linear.py:446]: 有的模型kv_proj也会用到MergedColumnParallelLinear

File: vllm_ascend/_310p/ops/linear.py [hunk 10]
```diff
@@ -0,0 +1,571 @@
...
+            output_size=int(part_local),
+            input_size=in_real,
+            output_offset=int(out_off),
+            tp_rank=self.tp_rank,
+            tp_size=self.tp_size,
+            prefix=self.prefix,
+        )
+
+
+class AscendReplicatedLinear310(ReplicatedLinear):
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/linear.py:502]: ReplicatedLinear用于MOE的门控层，qwen3-next的非TP并行层，以及VL模型的proj层，此处做padding会导致模型上下文维度出现异常。

File: vllm_ascend/_310p/ops/linear.py [hunk 11]
```diff
@@ -0,0 +1,571 @@
...
+        if local <= param_data.numel():
+            start = int(tp_rank) * int(local)
+            param_data.zero_()
+            param_data[:local].copy_(loaded.narrow(0, start, int(local)))
+        return True
+
+    raise RuntimeError(f"1D param mismatch: param={tuple(param_data.shape)} loaded={tuple(loaded.shape)}")
+
+
+class AscendUnquantizedLinearMethod310(UnquantizedLinearMethod):
```
Comment (WeNeedMoreCode) [inline vllm_ascend/_310p/ops/linear.py:129]: 为何310P需要单独新建UnquantizedLinearMethod，与其他设备有何不同

File: vllm_ascend/_310p/ops/activation.py [hunk 12]
```diff
@@ -16,13 +16,20 @@
...
+
+        out = torch_npu.npu_swiglu(x)
```
Comment (pu-zhe) [inline vllm_ascend/_310p/ops/activation.py:31]: 此处的融合算子性能收益很低，在300I DUO上E2E不足1%。并且可以做判断，
```
if x.shape[-1] // 32 == 0:
    out = torch_npu.npu_swiglu(x)
else:
    out = (F.silu(x[..., :h]) * x[..., h:])
```

**Comment by wangxiyuan** [comment]
no update for long time, close this now. Feel free to reopen if it's still needed.
